### DISCUSSION TITLE
Pagination for multiple components

### DISCUSSION THREAD

**Discussion by Dimonka2** [livewire/livewire#3145] [7. Bugs (archived)]
The issue is that Livewire paginator is using query string "page" and ignoring the value of `$pageName` passed to the Builder method `paginate()`.
Basically it is possible to have 2 Livewire components with pagination, but the browser query string will have only one `page` element reflecting last clicked page from any of those 2 components. 
Ideally Livewire paginator should keep the naming from the Builder paginator or at least to have some property or class constant in the component to adjust the name of the query string.
Thanks! 

- Livewire version: any
- Laravel version: latest
- Browser: any

**Comment by devcircus** [comment]
Noticed this myself. The pagination works correctly, it just doesn't reflect the passed path and only shows one "page" param. 

Thinking you can override the pagination methods and utilize events and query string updates to get what we need. I'll have to test it out.

**Comment by Dimonka2** [comment]
I have managed to make a trait, where your can redefine the query string for paginator.
It does not helps much with the same type of components on a page, but at least you can put two different components with paginators, Feel free to use it:

```php
<?php

namespace dimonka2\flatform\Traits;

use Illuminate\Pagination\Paginator;

trait WithPagination
{
    protected $pageName = 'page';

    public function getUpdatesQueryString()
    {
        $query = array_merge([$this->pageName => ['except' => 1]], $this->updatesQueryString);
        return $query;
    }

    public function initializeWithPagination()
    {
        $this->{$this->pageName} = $this->resolvePage();

        Paginator::currentPageResolver(function () {
            return $this->{$this->pageName};
        });

        Paginator::defaultView($this->paginationView());
    }

    public function paginationView()
    {
        return 'livewire::pagination-links';
    }

    public function previousPage()
    {
        $this->{$this->pageName} = $this->{$this->pageName} - 1;
    }

    public function nextPage()
    {
        $this->{$this->pageName} = $this->{$this->pageName} + 1;
    }

    public function gotoPage($page)
    {
        $this->{$this->pageName} = $page;
    }

    public function resetPage()
    {
        $this->{$this->pageName} = 1;
    }

    public function resolvePage()
    {
        // The "page" query string item should only be available
        // from within the original component mount run.
        return request()->query($this->pageName, $this->{$this->pageName});
    }

    public function __get($property) {
        if ($property == $this->pageName || $property == 'page') {
            $page = property_exists($this, $this->pageName) ? $this->{$this->pageName} : 1;
            return $page;
        }
      }

      public function __set($property, $value) {
        if ($property == $this->pageName || $property == 'page') {
            $this->{$this->pageName} = $value;
        }
        return $this;
      }

      public function getPublicPropertiesDefinedBySubClass()
      {
          $data = parent::getPublicPropertiesDefinedBySubClass();
          $data[$this->pageName] = $this->{$this->pageName};
          return $data;
      }
}
```
In your component obviously you need to define the property `$pageName`.

**Comment by Dimonka2** [comment]
There is an issue in my version of pagination - magic function `__get` needs to call parent `__get` function, otherwise should work.

**Comment by Wulfheart** [comment]
This is a pretty big issue.

**Comment by mazedlx** [comment]
This still exists in V2.2.5.

**Comment by jhoff** [comment]
I think I have a solid solution for this tied up. Some of this is informed by @Dimonka2's work above 🙌, but this doesn't require any changes to `__get`. It should be a drop in replacement for the stock `WithPagination` trait and should work with components that still use the stock trait as well.

It's worth noting that I've not yet tested this with Livewire v2, though it may work. I can submit it as a PR but it's probably introducing breaking changes, since it no longer relies on the public `page` property.

Create an `app/Http/Livewire/Traits/WithPagination.php` file:

```php
<?php

namespace App\Http\Livewire\Traits;

use Illuminate\Pagination\Paginator;

trait WithPagination
{
    protected $currentPage = 1;

    public function pageName()
    {
        return property_exists($this, 'pageName') ? $this->pageName : 'page';
    }

    public function getPageProperty()
    {
        return $this->currentPage;
    }

    public function getUpdatesQueryString()
    {
        return array_merge([$this->pageName() => ['except' => 1]], $this->updatesQueryString);
    }

    public function initializeWithPagination()
    {
        $this->currentPage = $this->resolvePage();

        Paginator::currentPageResolver(function () {
            return $this->currentPage;
        });

        Paginator::defaultView($this->paginationView());
    }

    public function paginationView()
    {
        return 'livewire::pagination-links';
    }

    public function previousPage()
    {
        $this->currentPage = $this->currentPage - 1;
    }

    public function nextPage()
    {
        $this->currentPage = $this->currentPage + 1;
    }

    public function gotoPage($page)
    {
        $this->currentPage = $page;
    }

    public function resetPage()
    {
        $this->currentPage = 1;
    }

    public function resolvePage()
    {
        return request()->query($this->pageName(), $this->currentPage);
    }

    public function getPublicPropertiesDefinedBySubClass()
    {
        return tap(parent::getPublicPropertiesDefinedBySubClass(), function (&$props) {
            $props[$this->pageName()] = $this->currentPage;
        });
    }
}
```

If you only use one paginated component on a page and just want to use the default page name of `page`, simply add this trait. No other modification needed.

```php
namespace App\Http\Livewire;

use Livewire\Component;
use App\Http\Livewire\Traits\WithPagination;

class Foobar extends Component
{
    use WithPagination;

    ...
}
```

If you have multiple paginated components on a page but they are all different component types, then you can simply add `public $pageName = 'foobarPage';` to each component to make the page name unique.

> Note: you don't have to customize all of the page names. If you want the main paginator to still use `page`, that'll work just fine.

```php
namespace App\Http\Livewire;

use Livewire\Component;
use App\Http\Livewire\Traits\WithPagination;

class Foobar extends Component
{
    use WithPagination;

    public $pageName = foobarPage';

    ...
}
```

However, if you have multiple paginated components on a page where two or more are of the same component type, then you have to define a `pageName` method in your component that returns a unique page name for each component instance.


```php
namespace App\Http\Livewire;

use Livewire\Component;
use App\Http\Livewire\Traits\WithPagination;

class Foobar extends Component
{
    use WithPagination;

    public $uniqueId;

    public function pageName()
    {
        return 'foobarPage' . $this->uniqueId;
    }

    public function mount($uniqueId)
    {
        $this->uniqueId = $uniqueId;
    }

    ...
}
```

It's worth noting that when you create your paginator, you probably want to set the page name there as well. This can be easily done by using `$this->pageName()` as the third argument.

```php
    public function getResultsProperty()
    {
        return Auth::user()->foobar()->paginate(null, ['*'], $this->pageName());
    }
```

Note that this step isn't technically required, as Livewire doesn't really care what the paginator thinks it's name should be, however the anchor tags will use the incorrect page name if you skip this step. If a user opens a page link in a new tab for example, it wouldn't send them to the expected page.

**Comment by Wulfheart** [comment]
I think this is something @calebporzio or @imliam should approve.

**Comment by erikwestlund** [comment]
I do think it's worth thinking about how to put in an official patch for this. In another issue, Caleb suggested maintaining one's own pagination trait. I've gotten that to work, sort of: my paginator didn't update the current page on page change. I guess something of a dom diffing issue.

I've taken to just not doing standard pagination on pages like this which is not ideal.

**Comment by jhoff** [comment]
@erikwestlund 

> my paginator didn't update the current page on page change

I think the version that posted should fix that for you, if I understand what you're saying.

I'm about 2 weeks out from upgrading my app to v2. When I do, I'll submit a PR for this with tests and everything.

**Comment by Wulfheart** [comment]
@jhoff Any news on this`?

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by ragingdave** [comment]
REOPEN

This seems like definitely something that needs to be addressed as without this change, you can't have multiple paginated components on a single page. This could be something for sure useful on more complex UI's with multiple lists or even tabbed content that contains multiple livewire components.

**Comment by martianatwork** [comment]
I have tested the code on v2, inspired by @Dimonka2 and @jhoff code.

```php
<?php

namespace App\Traits;
use Illuminate\Pagination\Paginator;
use Livewire\WithPagination;

trait CustomWithPagination
{
    use WithPagination;

    public function pageName(): string {
        if (property_exists($this, 'pageName')) {
            if(!isset($this->{$this->pageName})) $this->{$this->pageName} = 1;
            return $this->pageName;
        } else {
            return 'page';
        }
    }

    public function getQueryString()
    {
        return array_merge([$this->pageName() => ['except' => 1]], $this->queryString);
    }

    public function initializeWithPagination() {
        $this->{$this->pageName()} = $this->resolvePage();

        Paginator::currentPageResolver(function () {
            return $this->{$this->pageName()};
        });

        Paginator::defaultView($this->paginationView());
    }


    public function previousPage()
    {
        $this->setPage($this->{$this->pageName()} - 1);
    }

    public function nextPage()
    {
        $this->setPage($this->{$this->pageName()} + 1);
    }

    public function setPage($page)
    {
        $this->{$this->pageName()} = $page;
    }

    public function resolvePage() {
        return request()->query($this->pageName(), $this->{$this->pageName()});
    }

    public function getPublicPropertiesDefinedBySubClass() {
        return tap(parent::getPublicPropertiesDefinedBySubClass(), function (&$props) {
            $props[$this->pageName()] = $this->{$this->pageName()};
        });
    }
}
```

in your Component add

```php
    public $pageName = 'settledPageName';
```

I'm extending the `WithPagination` as it provides basic code such as paginationView, etc. also i feel like if the base is updated then we dont have to update the CustomPagination unless it is a big change.

I also had to update the pagination query
```php
->paginate(10, ['*'], $this->pageName())
```

**Comment by ragingdave** [comment]
@martianatwork While I appreciate the suggestion and the code snippets from previous suggestions, I would argue that this or a similar approach should be handled in livewire directly as this seems like it would be a fairly common occurrence and shouldn't need external overriding to make work. I also think that it should be some sort of automatic page name, so that a dev doesn't have to specify the page name to use manually.

@calebporzio maybe give this some more thought?

**Comment by martianatwork** [comment]
@ragingdave Yes, you are right.

This should be based on components class name minus the component part if it is a part of the name.

i think we should work on the 
```php
->paginate(10, ['*'], $this->pageName())
```

Part as this is very annoying, we can create a `paginate` function which accepts same parameters as eloquent's pagination.

**Comment by Hrituraj202** [comment]
@martianatwork this works great with multiple components on the same page but fails in case of multiple paginations on the same component

**Comment by camya** [comment]
REOPEN

I also struggled with this problem quite some times today. After trying, and investigating the vendor folder, and searching the web, ... I found this closed thread. I hope, @calebporzio can reopen it. 

My demo page contains a simple list of post links next to a list of the corresponding post comments. Both lists are pageable. 

**Required feature:** Possibility to set a custom page parameter name in Livewire. (e.g. 'page-posts' and a 'page-post-comments' instead of 'page')

**Feature request:** Allow to **remove the "page" parameter** for a Livewire component completely for Ajax only pagination.

In my example I normally really don't need the page parameter for the post-comments in the URL. Users should not bookmark a paged comment page in my case. After removing it, the pagination is done without modifyng the URL.

**Comment by ninthspace** [comment]
I've worked around this issue by retaining all criteria for multiple paginated components in the user's session identified by a unique id, and just using that unique id in the query string which is passed to all components in a view. Obviously it means that whenever criteria get updated the session needs to be modified, but storing criteria this way allows me to use Livewire's `InteractsWithProperties::fill` method to populate a component when it is mounted. Because the initial state of a page is then derived from that unique id, it retains any existing pagination (and anything else I choose to retain) when navigating backwards and forwards through browser history.

**Comment by Whallas** [comment]
From what I could see in the community this is a necessary feature. I had the same problem a few days ago, I was thinking of creating a pull request but I found this issue enough and the suggestions here are very good.

**Comment by Wulfheart** [comment]
@Whallas Please create a PR so that it is supported out of the box.  🙏

**Comment by madstone-dev** [comment]
I simply solved this

```
public function isPaginateEnabled()
{
    if (property_exists($this, 'usePagination')) {
        return $this->usePagination;
    }
    return true;
}

public function getQueryString()
{
    if($this->isPaginateEnabled()){
        return array_merge(['page' => ['except' => 1]], $this->queryString);
    }
}
```

Put this code in the With Pagination.php in the vendor folder

```
protected $usePagination = false;
```

Then enter this code in the component and you're done.


Sorry, it seems unrelated to this issue.
I just tried to pagination a component without updating the url in a single page.

**Comment by ragingdave** [comment]
@madstone-dev That's not really a solution because it doesn't look like it would solve the issue of multiple components on a single page for pagination on all of them. Additionally, that requires vendor level changes which is a HUGE no-no as that hinders upgradability. I'm not entirely sure what you are attempting to fix though as this code just seems to disable/enable pagination entirely....even...if...it's...included with the trait? Which isn't really the issue outlined in this issue.

**Comment by andreasmatuszewski** [comment]
I would love to see this feature

**Comment by ariaieboy** [comment]
I add a PR you guys can review it
https://github.com/livewire/livewire/pull/3264

**Comment by kwienken** [comment]
I also ran into this pagination and query string problem when rendering the same Livewire table component multiple times with different data on the same page. I resolved this a bit differently though with minimal changes to the component class and without a new pagination trait. 

My tables are paginated with sorting on a single column. The query string keys for sort direction and page needed to be unique for each rendered table component. 

First I added a public query string index property to the table component and set it when including.

```
/**
 * @var int
 */
public $qsIndex;
```

`
<livewire:match.result.table :qs-index="$index" :match="$match" />
`

I then updated the `mount` method on the table component to set the sorting and pagination properties from the query string.
```
/**
 * @param Request $request
 * @param Match $match
 * @param int $qsIndex
 * @return void
 */
public function mount(Request $request, Match $match, int $qsIndex)
{
    $this->match = $match;
    $this->qsIndex = $qsIndex;

    $sortDirectionAlias = $this->getSortDirectionAlias();
    if ($request->has($sortDirectionAlias)) {
        $this->sortDirection = $request->query($sortDirectionAlias);
    }

    $pageAlias = $this->getPageAlias();
    if ($request->has($pageAlias)) {
        $this->setPage($request->query($pageAlias));
    }
}
```
I used the `setPage` method from the vendor `WithPagination` trait because pagination is initialized before the `mount` method executes.

Here are the get methods for the query string aliases.
```
/**
 * Gets the sort direction query string alias
 *
 * @return string
 */
protected function getSortDirectionAlias(): string
{
    return "sd{$this->qsIndex}";
}

/**
 * Gets the sort direction query string alias
 *
 * @return string
 */
protected function getPageAlias(): string
{
    return "pg{$this->qsIndex}";
}
```

Finally I added a `queryString()` method to override the page pagination property with the unique page alias and add the aliased sort direction property.
```
/**
 * Gets the query string definition
 *
 * @return array[]
 */
protected function queryString(): array
{
    return [
        'sortDirection' => [
            'except' => 'desc',
            'as' => $this->getSortDirectionAlias(),
        ],
        'page' => [
            'except' => 1,
            'as' => $this->getPageAlias(),
        ],
    ];
}
```

I hope this helps others that may stumble upon this discussion. :)

**Reply by joshhanley** [reply]
@kwienken thanks for sharing your solution!

We did end up adding support for multiple paginators in v2.6 see PR #3684. Does that work for your usecase?
